nixos/networkd: update configuration options

andir opened the pull request:
##### Motivation for this change

While working on my person home router configuration I started adding a few more options to our systemd-networkd configuration options. The entire area around prefix delegation wasn't really covered. Also upstream just recently added support for `PrefixDelegationHint` which is a must-have  (for me). That flag will become available soon. @NinjaTrappeur is working on the systemd v244 upgrade.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

andir committed:
nixos/networkd: add RoutesToDNS to DHCP section

andir committed:
nixos/networkd: remove CriticalConnection= fields in favor of KeepConnection

Systemd upstream has deprecated CriticalConnection with v244 in favor of
KeepConnection as that seems to be more flexible:

  The CriticalConnection= setting in .network files is now deprecated,
  and replaced by a new KeepConfiguration= setting which allows more
  detailed configuration of the IP configuration to keep in place.

andir committed:
nixos/networkd: add MaxAttempts and SendRelease to the DHCP options

andir committed:
nixos/networkd: introduce the DHCPv6 network section eqivalent

You can now specify option for the `[DHCPv6]` section with
`systemd.network.<name>.dhcpV6Config.…`. Previously you could only use
the combined legacy DHCP configuration.

andir committed:
nixos/networkd: rename the networkd dhcpConfig option to dhcpV4Config

This follows upstreams change in documentation. While the `[DHCP]`
section might still work it is undocumented and we should probably not
be using it anymore. Users can just upgrade to the new option without
much hassle.

I had to create a bit of custom module deprecation code since the usual
approach doesn't support wildcards in the path.

andir committed:
nixos/networkd: add ipv6PrefixDelegationConfig to networkd

andir committed:
nixos/networkd: add ipv6Prefix

andir committed:
nixos/networkd: support PrefixDelegationHint in DHCPv6 section

With sytemd v244 we will have support for this option.

NinjaTrappeur commented:
Awesome, it was on my todo list!

One thing though: the systemd configuration parser won't error when encountering some unknown keys. That's kind of an issue when we'll update systemd and a key has been renamed. The module could silently break.

One way to mitigate that problem is to add some nixos tests and check that each networkd directive has indeed been applied to a test network. We started doing that in nixos/tests/systemd-networkd.nix.

Do you plan to add some tests for the RA/DHCPv6 setup? If you don't, could you please wait for me to add some RA-related tests before merging that?

Ma27 reviewed (approved):
LGTM :+1:

flokli commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -72,6 +72,16 @@
+     <para>
+       The <literal>systemd-networkd</literal> option
+       <literal>systemd.network.networks.&lt;name&gt;.dhcp.CriticalConnection</literal>
+       has been removed following upstream systemds deprecation of the same. It is recommended to use
```
```suggestion
       has been removed following upstream systemd's deprecation of the same. It is recommended to use
```

flokli commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -277,7 +278,7 @@ let
     (assertOnlyFields [
       "UseDNS" "RoutesToDNS" "UseNTP" "UseMTU" "Anonymize" "SendHostname" "UseHostname"
-      "Hostname" "UseDomains" "UseRoutes" "UseTimezone" "CriticalConnection"
+      "Hostname" "UseDomains" "UseRoutes" "UseTimezone"
```
Upstream removed it from their docs, but they still allow setting it to keep compatibility.

We will now fail, and force users to migrate to the new setting - which might be the better decision, given we can have these checks at configuration time, and don't need to keep supporting old field indefinitely, as unknown fields are just ignored upstream.

flokli commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -641,13 +640,13 @@ let
     };
 
-    dhcpConfig = mkOption {
+    dhcpV4Config = mkOption {
```
We probably still want to add a `mkRemovedOptionModule`, and print a short summary of what's in the release notes instead of just dropping the configuration options here.

andir replied:
I tried adding that but due to the nature of how those work that isn't really feasible here. We have a "wildcard" network name in between and can only really assert if someone tries to access that value.

I tried a few things:

- traversing the `config` tree to see if the option is set, that does then trigger the "on access" assertion
- traversing the `option` tree (as it is done in the original helper) but that doesn't seem possible for these kinds of situations..

Here is the code I ended up backing out of this changest in case anyone wants to give it a shot:

```nix
# systemd.network.networks.*.dhcpConfig has been deprecated in favor of ….dhcpV4Config
# Produce a nice warning message so users know it is gone.
{
  assertions = let
    #affectedNetworks = attrNames (filterAttrs (network: conf: conf.dhcpConfig.isDefined) options.systemd.network.networks.values);
    affectedNetworks = attrNames (filterAttrs (network: conf: conf ? dhcpConfig) config.systemd.network.networks);
  in
  [
    {
      assertion = length affectedNetworks == 0;
      message = ''
        The option definition `systemd.network.networks.*.dhcpConfig` has been deprecated in favor of `systemd.network.networks.*.dhcpV4Config`. Please update your configuration.
        The following networks are still using it:
        ${concatMapStringsSep "\n" (network: "  - `systemd.network.networks.\"${network}\"`") affectedNetworks}
      '';
    }
  ];
}

```

andir replied:
For now I just left the "on access" throw in the code as that is the least we could do. I do not really want to make it an alias.

flokli commented on nixos/modules/system/boot/networkd.nix:
```diff
@@ -303,6 +303,7 @@ let
   checkDhcpV6 = checkUnitConfig "DHCPv6" [
     (assertOnlyFields [
       "UseDns" "UseNTP" "RapidCommit" "ForceDHCPv6PDOtherInformation"
+      "PrefixDelegationHint"
```
Withs->With in the commit message. Also, depending on timing, we should not merge this commit before the actual bump to systemd 244 or 245 has happened. Otherwise, users might be confused about using a option that's exposed in the NixOS module system, but silently ignored by systemd.

I'd propose fixing the other parts of this PR and dropping this commit, to not block it on the version bumps.

flokli commented:
@andir can you address the requested changes, and rebase on latest master?

andir commented:
On 00:49 08.03.20, Félix Baylac-Jacqué wrote:
> Awesome, it was on my todo list!
> 
> One thing though: the systemd configuration parser won't error when
> encountering some unknown keys. That's kind of an issue when we'll
> update systemd and a key has been renamed. The module could silently
> break.
> 
> One way to mitigate that problem is to add some nixos tests and check
> that each networkd directive has indeed been applied to a test
> network. We started doing that in nixos/tests/systemd-networkd.nix.
> 
> Do you plan to add some tests for the RA/DHCPv6 setup? If you don't,
> could you please wait for me to add some RA-related tests before
> merging that?

Yes, I have some extensive test suite locally that tests the whole
IA-NA/IA-PD setup but I do not intend to commit it with this PR. It
needs a bunch of polishing. Will likely happen soon(tm) but not
immediately.

andir commented:
Rebased and fixed a few typos. Any further requests? @flokli 

As I stated earlier the tests I plan to add are probably too big to be in scope of this change and will require some discussions while these changes here should not be controversial.

flokli commented:
Adding the tests later is fine. Let's merge this :+1:
